### PR TITLE
tokenWithAuthorizedEntity: method should also check if token is fresh

### PR DESCRIPTION
**Pull request by chliangGoogle**
This API is used to fetch token directly from server. Right now if APNS info is not changed, the client will simply use cached token.
We should also check token freshness, which checks if IID and token is consistent or if app version has changed, etc. This helps us migrate to FIS. (b/140506963)

### CODE REVIEW

**Comment by chliangGoogle** [comment]
@maksymmalyhin This change needs to be production as well because we don't want APNS info block any token update. I can check in to your branch after this PR is submitted.

File: Firebase/InstanceID/FIRInstanceIDTokenManager.m [hunk 1]
```diff
@@ -281,12 +281,12 @@ - (BOOL)checkTokenRefreshPolicyWithIID:(NSString *)IID {
   NSMutableArray<FIRInstanceIDTokenInfo *> *tokenInfosToDelete =
       [NSMutableArray arrayWithCapacity:tokenInfos.count];
   for (FIRInstanceIDTokenInfo *tokenInfo in tokenInfos) {
-    BOOL isTokenFresh = [tokenInfo isFresh];
-    if (isTokenFresh && [tokenInfo.token hasPrefix:IID]) {
+    BOOL isTokenFresh = [tokenInfo isFreshWithIID:IID];
+    if (isTokenFresh) {
```
Comment (maksymmalyhin) [inline Firebase/InstanceID/FIRInstanceIDTokenManager.m]: nit: Doesn't look like the variable is used anywhere else. If so, can we just call the method directly in `if`?

File: Example/InstanceID/Tests/FIRInstanceIDTokenInfoTest.m [hunk 2]
```diff
@@ -126,55 +127,67 @@ - (void)testTokenInfoEncodingAndDecodingWithMissingFields {
...
       [[FIRInstanceIDTokenInfo alloc] initWithAuthorizedEntity:kAuthorizedEntity
                                                          scope:kScope
                                                          token:kToken
                                                     appVersion:@"1.1"
                                                  firebaseAppID:FIRInstanceIDFirebaseAppID()];
-  XCTAssertFalse([self.validTokenInfo isFresh]);
+  XCTAssertFalse([self.validTokenInfo isFreshWithIID:kIID]);
+}
+
+-(void)testTokenInconsistentWithToken {
```
Comment (maksymmalyhin) [inline Example/InstanceID/Tests/FIRInstanceIDTokenInfoTest.m]: Did you mean `testTokenInconsistentWithIID`?

File: Firebase/InstanceID/FIRInstanceIDTokenInfo.h [hunk 3]
```diff
@@ -74,8 +74,18 @@ NS_ASSUME_NONNULL_BEGIN
...
+ * 5. token is consistent with the current IID.
+ * 6. APNS info has changed.
+ * @param IID  The app identifiier that is used to check if token is prefixed with.
+ * @return If token is fresh.
+ *
+ */
+- (BOOL)isFreshWithIID:(NSString *)IID;
+
+/*
+ * Check wehther the token is default token.
```
Comment (morganchen12) [inline Firebase/InstanceID/FIRInstanceIDTokenInfo.h]: nit: `whether` typo

Comment (chliangGoogle) [reply Firebase/InstanceID/FIRInstanceIDTokenInfo.h]: Done! Thanks!

**Comment by maksymmalyhin** [comment]
@chliangGoogle As for the functional test - I think we may avoid using different versions of IID for this specific case. I was thinking of mocking the token manager (or even putting the storage into the state) so InstanceID is not consistent with the token, then call the method and make sure all required methods called and eventual result is correct.

**Comment by chliangGoogle** [comment]
Yep, that's exactly what new test I added is doing.

**Comment by chliangGoogle** [comment]
This should fix https://github.com/firebase/firebase-ios-sdk/issues/4167

**Pull request merged**